engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+          return undefinedValue;
+        }
+        String strToMatch = (String) ArgList[0];
+        Pattern p = Pattern.compile( (String) ArgList[1] );
+        Matcher m = p.matcher( strToMatch );
+        if ( m.matches() && m.groupCount() > 0 ) {
+          strArr = new String[m.groupCount()];
+          for ( int i = 1; i <= m.groupCount(); i++ ) {
+            strArr[i - 1] = m.group( i );
+          }
+        }
+      } catch ( Exception e ) {
+        throw new RuntimeException( e.toString() );
+      }
+    } else {
+      throw new RuntimeException( "The function call str2RegExp requires 2 arguments." );
+    }
+    return strArr;
+  }
+
+  public static void touch( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      if ( ArgList.length == 1 && !isNull( ArgList[0] ) && !isUndefined( ArgList[0] ) ) {
+        File file = new File( (String) ArgList[0] );
+        boolean success = file.createNewFile();
+        if ( !success ) {
+          file.setLastModified( System.currentTimeMillis() );
+        }
+      } else {
+        throw new RuntimeException( "The function call touch requires 1 valid argument." );
+      }
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+  }
+
+  public static Object fileExists( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      if ( ArgList.length == 1 && !isNull( ArgList[0] ) && !isUndefined( ArgList[0] ) ) {
+        if ( ArgList[0].equals( null ) ) {
+          return null;
+        }
+        File file = new File( (String) ArgList[0] );
+        return Boolean.valueOf( file.isFile() );
+      } else {
+        throw new RuntimeException( "The function call fileExists requires 1 valid argument." );
+      }
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+  }
+
+  public static Object str2date( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    Object oRC = new Object();
+    String sArg1 = "";
+    String sArg2 = "";
+    String sArg3 = "";
+    String sArg4 = "";
+    switch ( ArgList.length ) {
+      case 0:
+        throw new RuntimeException( "Please provide a valid string to the function call str2date." );
+      case 1:
+        try {
+          if ( isNull( ArgList[0] ) ) {
+            return null;
+          } else if ( isUndefined( ArgList[0] ) ) {
+            return undefinedValue;
+          }
+          sArg1 = (String) ArgList[0];
+          Format dfFormatter = new SimpleDateFormat();
+          oRC = dfFormatter.parseObject( sArg1 );
+          // if(Double.isNaN(sArg1)) throw new RuntimeException("The first Argument must be a Number.");
+          // DecimalFormat formatter = new DecimalFormat();
+          // sRC= formatter.format(sArg1);
+        } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 1030, 1053, 1083.